sdks/java/io/kafka/upgrade/src/main/java/org/apache/beam/sdk/io/kafka/upgrade/KafkaIOTranslation.java

```diff
@@ -203,162 +210,195 @@ public Row toConfigRow(Read<?, ?> transform) {
-      }
-      Instant stopReadTime = configRow.getValue("stop_read_time");
-      if (stopReadTime != null) {
-        transform = transform.withStopReadTime(stopReadTime);
-      }
-      Boolean isCommitOffsetFinalizeEnabled =
-          configRow.getBoolean("is_commit_offset_finalize_enabled");
-      if (isCommitOffsetFinalizeEnabled != null && isCommitOffsetFinalizeEnabled) {
-        transform = transform.commitOffsetsInFinalize();
-      }
-      Boolean isDynamicRead = configRow.getBoolean("is_dynamic_read");
-      if (isDynamicRead != null && isDynamicRead) {
-        Duration watchTopicPartitionDuration = configRow.getValue("watch_topic_partition_duration");
-        if (watchTopicPartitionDuration == null) {
-          throw new IllegalArgumentException(
-              "Expected watchTopicPartitionDuration to be available when isDynamicRead is set to true");
+        byte[] checkStopReadinfFn = configRow.getBytes("check_stop_reading_fn");
+        if (checkStopReadinfFn != null) {
+          transform =
+              transform.withCheckStopReadingFn(
+                  (SerializableFunction<TopicPartition, Boolean>)
+                      fromByteArray(checkStopReadinfFn));
         }
-        transform =
-            transform.withDynamicRead(
-                org.joda.time.Duration.millis(watchTopicPartitionDuration.toMillis()));
-      }
 
-      byte[] timestampPolicyFactory = configRow.getBytes("timestamp_policy_factory");
-      if (timestampPolicyFactory != null) {
-        transform =
-            transform.withTimestampPolicyFactory(
-                (TimestampPolicyFactory) fromByteArray(timestampPolicyFactory));
-      }
-      Map<String, byte[]> offsetConsumerConfig = configRow.getMap("offset_consumer_config");
-      if (offsetConsumerConfig != null) {
-        Map<String, Object> updatedOffsetConsumerConfig = new HashMap<>();
-        offsetConsumerConfig.forEach(
-            (key, dataBytes) -> {
-              if (offsetConsumerConfig.get(key) == null) {
-                throw new IllegalArgumentException(
-                    "Encoded value for the offset consumer config key " + key + " was null.");
-              }
-              updatedOffsetConsumerConfig.put(key, fromByteArray(offsetConsumerConfig.get(key)));
-            });
-        transform = transform.withOffsetConsumerConfigOverrides(updatedOffsetConsumerConfig);
-      }
+        byte[] badRecordErrorHandlerBytes = configRow.getBytes("bad_record_error_handler");
+        if (badRecordErrorHandlerBytes != null) {
+          try {
+            transform =
+                transform.withBadRecordErrorHandler(
+                    (ErrorHandler) fromByteArray(badRecordErrorHandlerBytes));
+          } catch (InvalidClassException e) {
+            LOG.warn(
+                "Could not use the provided `ErrorHandler` implementation when upgrading."
+                    + "Using the default.");
+          }
```

Comment:
the transform behavior changes whether or not an error handler is passed at all, so this could break if deserialization fails.

Reply:
I think ErrorHandler has to be instantiated with it's own schema but will do this in a separate PR.
For now, updating the code so that we do a hard fail if the user set this property.